Correct `DownloadTracker` reporting on retries and failures

FranciscoTGouveia opened the pull request:
After refactoring the `DownloadTracker` to use `indicatif` and enabling concurrent downloads, a few *UX* issues arose.

This PR addresses those issues, specifically fixing the cluttered output that occurred whenever a download was retried or failed.

An example of the cluttered output described above is shown below:

<img width="793" height="683" alt="image" src="https://github.com/user-attachments/assets/9f5b6a42-f354-4cdd-8db8-71579d7d3662" />

FranciscoTGouveia committed:
fix(downloads): correct faulty output when retrying a download

The `DownloadFinished` notification was always being called even if the
download was unsuccessful. This would confuse the DownloadTracker as to
the download really ended (successfully).

To fix this, the notification is now only thrown iff the download didn't
timeout. This could be (possibly) accompanied by a new notification, say
`DownloadTimedOut` but it wouldn't have any direct use yet, so I refrain
from adding it.

FranciscoTGouveia committed:
fix(downloads): correct faulty behavior when a download fails

Until now, the progress bars were not being correctly handled which
would make an error (i.e. "failed to download file X") pollute the
output.

rami3l commented on src/cli/download_tracker.rs:
```diff
@@ -105,4 +112,30 @@ impl DownloadTracker {
+        };
+        pb.set_style(
+            ProgressStyle::with_template(
+                "{msg:>12.bold}  [{bar:40}] {bytes}/{total_bytes} ({bytes_per_sec}, ETA: {eta})",
```
Hmmm this looks reasonable impl-wise but I'm afraid the user will not be very clear of what will happen after, say, `download failed after 2s`. Can we somehow delay the reappearance of the progress bar to a later notification, if possible, and saying something like `retrying download` here instead?

FranciscoTGouveia replied:
Thanks for the comment!

Regarding the part highlighted in this comment, you’re right -- it should explicitly inform the user that the download is being retried. I’ll fix that.
My initial _rationale_ was that, since the progress bar already resets to the beginning, it would be clear that the download was starting over.

As for delaying the reappearance of the progress bar to give users time to see the retrying download message, I’ll look into that now.

FranciscoTGouveia replied:
I’ve already addressed this.
The reappearance of the `ProgressBar` is now delayed by 1 second -- do you think that’s sufficient?

rami3l replied:
@FranciscoTGouveia Yes, it looks good to me this time, thanks again :)

rami3l reviewed (changes requested):
Many thanks for this PR :)

djc commented on src/download/mod.rs:
```diff
@@ -218,7 +218,12 @@ async fn download_file_(
+        notify_handler(Notification::DownloadFinished(Some(url.as_str())));
+    } else {
+        notify_handler(Notification::DownloadFailed(url.as_str()));
+    }
```
Nit: suggest rewriting this as

```rust
notify_handler(match &res {
     Ok(_) => Notification::DownloadFinished(Some(url.as_str()))),
     Err(_) => Notification::DownloadFailed(url.as_str()),
})
```

FranciscoTGouveia replied:
Thanks!

FranciscoTGouveia committed:
feat(downloads): delay the reappearance of the progress bar when retrying a download

To achieve this, an `Instant` was coupled with the actual ProgressBar.
This allow us to only trigger the progress bar after 1 second has passed
allowing the user to see that the download is being retried.

rami3l commented on src/cli/download_tracker.rs:
```diff
@@ -14,7 +15,7 @@ pub(crate) struct DownloadTracker {
     /// MultiProgress bar for the downloads.
     multi_progress_bars: MultiProgress,
     /// Mapping of URLs being downloaded to their corresponding progress bars.
```
You'll need to update the docstring as well :)

rami3l replied:
@FranciscoTGouveia In addition to what it does, you should also write what it is which is equally important. Something similar to "The `Option<T>` represents the instant of the last retry..." but in your own words.

rami3l reviewed (approved):
LGTM modulo a tiny nit :)

djc commented on src/cli/download_tracker.rs:
```diff
@@ -83,27 +84,40 @@ impl DownloadTracker {
+        pb.inc(len as u64);
+        if let Some(retry_instant) = retry_time
+            && retry_instant.elapsed() > Duration::from_secs(1)
+        {
```
Invert this to reduce rightward drift:

```rust
match retry_time {
     Some(retry_time) if retry_instant.elapsed() > Duration::from_secs(1) => {}
     _ => return,
}
```

rami3l replied:
@FranciscoTGouveia I think you have misunderstood the comment above. Since `retry_time` is not really used, either the one suggested above or `if !retry_time.is_some_and(..)` with early return will be clearer.

FranciscoTGouveia replied:
Sorry about that! :sweat_smile: 
I’ve already addressed it, opting for an early return.
